provider/pkg/metadata/naming_test.go

```diff
@@ -81,11 +82,33 @@ func TestAssignNameIfAutonamable(t *testing.T) {
 		}),
 	}
 	o5 := propMapToUnstructured(pm5)
-	AssignNameIfAutonamable(nil, o5, pm5, resource.NewURN(tokens.QName("teststack"), tokens.PackageName("testproj"),
+	AssignNameIfAutonamable(nil, nil, o5, pm5, resource.NewURN(tokens.QName("teststack"), tokens.PackageName("testproj"),
 		tokens.Type(""), tokens.Type("bang:boom/fizzle:MajorResource"), "foo"))
 	assert.False(t, IsAutonamed(o5))
 	assert.Equal(t, "", o5.GetGenerateName())
 	assert.Equal(t, "", o5.GetName())
+
+	// o6 has no name, a name is proposed by the engine, so autonaming picks the proposed name.
+	o6 := &unstructured.Unstructured{}
+	autonamingProposed := &pulumirpc.CheckRequest_AutonamingOptions{
+		Mode:         pulumirpc.CheckRequest_AutonamingOptions_PROPOSE,
+		ProposedName: "bar",
+	}
+	AssignNameIfAutonamable(nil, autonamingProposed, o6, pm1, resource.NewURN(tokens.QName("teststack"), tokens.PackageName("testproj"),
+		tokens.Type(""), tokens.Type("bang:boom/fizzle:MajorResource"), "foo"))
+	assert.True(t, IsAutonamed(o6))
+	assert.Equal(t, "bar", o6.GetName())
+
+	// o7 has no name, but autonaming is disabled by the engine, so autonaming fails.
+	o7 := &unstructured.Unstructured{}
+	autonamingDisabled := &pulumirpc.CheckRequest_AutonamingOptions{
+		Mode: pulumirpc.CheckRequest_AutonamingOptions_DISABLE,
+	}
+	AssignNameIfAutonamable(nil, autonamingDisabled, o7, pm1, resource.NewURN(tokens.QName("teststack"), tokens.PackageName("testproj"),
+		tokens.Type(""), tokens.Type("bang:boom/fizzle:MajorResource"), "foo"))
+	assert.False(t, IsAutonamed(o7))
+	assert.Equal(t, "", o7.GetGenerateName())
+	assert.Equal(t, "", o7.GetName())
```

Comment:
In this case autonaming doesn't _fail_, the object just doesn't get a name. That will ultimately become an error when we try to apply the object to the cluster, since the name is required.

If we encounter this situation we should return an error to the user instead of letting them apply the invalid object. Something like "Autonaming is disabled so XYZ requires a .metadata.name". It would make sense to modify `AssignNameIfAutonamable` to return `error` for this situation.

Reply:
Is it always an error if a `name` is not specified, for all resource types? I assumed that some objects may allow that? But it sounds like that's not true? If name/generateName are always required, an error makes sense to me.

Reply:
Added an error on Disabled